src/crypto/crypto_util.h

```diff
@@ -0,0 +1,680 @@
+  // any allocated data. The ByteSource will be reset with size = 0
+  // after being called.
+  std::unique_ptr<v8::BackingStore> ReleaseToBackingStore();
+
+  v8::Local<v8::ArrayBuffer> ToArrayBuffer(Environment* env);
+
+  void reset();
+
+  // Allows an Allocated ByteSource to be truncated.
+  inline void Resize(size_t newsize) {
+    CHECK_LE(newsize, size_);
+    CHECK_NOT_NULL(allocated_data_);
+    char* new_data_ = ReallocOpenSSL<char>(allocated_data_, newsize);
+    data_ = allocated_data_ = new_data_;
+    size_ = newsize;
+  }
+
+  static ByteSource Allocated(char* data, size_t size);
+  static ByteSource Foreign(const char* data, size_t size);
+
+  static ByteSource FromEncodedString(Environment* env,
+                                      v8::Local<v8::String> value,
+                                      enum encoding enc = BASE64);
+
+  static ByteSource FromStringOrBuffer(Environment* env,
+                                       v8::Local<v8::Value> value);
+
+  static ByteSource FromString(Environment* env,
+                               v8::Local<v8::String> str,
+                               bool ntc = false);
+
+  static ByteSource FromBuffer(v8::Local<v8::Value> buffer,
+                               bool ntc = false);
+
+  static ByteSource FromBIO(const BIOPointer& bio);
+
+  static ByteSource NullTerminatedCopy(Environment* env,
+                                       v8::Local<v8::Value> value);
+
+  static ByteSource FromSymmetricKeyObjectHandle(v8::Local<v8::Value> handle);
+
+  ByteSource(const ByteSource&) = delete;
+  ByteSource& operator=(const ByteSource&) = delete;
+
+  static ByteSource FromSecretKeyBytes(
+      Environment* env, v8::Local<v8::Value> value);
+
+ private:
+  const char* data_ = nullptr;
+  char* allocated_data_ = nullptr;
+  size_t size_ = 0;
+
+  ByteSource(const char* data, char* allocated_data, size_t size);
+};
+
+enum CryptoJobMode {
+  kCryptoJobAsync,
+  kCryptoJobSync
+};
+
+inline CryptoJobMode GetCryptoJobMode(v8::Local<v8::Value> args) {
+  CHECK(args->IsUint32());
+  uint32_t mode = args.As<v8::Uint32>()->Value();
+  CHECK_LE(mode, kCryptoJobSync);
+  return static_cast<CryptoJobMode>(mode);
+}
+
+template <typename CryptoJobTraits>
+class CryptoJob : public AsyncWrap, public ThreadPoolWork {
+ public:
+  using AdditionalParams = typename CryptoJobTraits::AdditionalParameters;
+
+  inline explicit CryptoJob(
+      Environment* env,
+      v8::Local<v8::Object> object,
+      AsyncWrap::ProviderType type,
+      CryptoJobMode mode,
+      AdditionalParams&& params)
+      : AsyncWrap(env, object, type),
+        ThreadPoolWork(env),
+        mode_(mode),
+        params_(std::move(params)) {
+    // If the CryptoJob is async, then the instance will be
+    // cleaned up when AfterThreadPoolWork is called.
+    if (mode == kCryptoJobSync) MakeWeak();
+  }
+
+  inline void AfterThreadPoolWork(int status) override {
```

Comment:
```suggestion
  void AfterThreadPoolWork(int status) override {
```

(I’ll stop commenting this now :smile: Basically, `inline` for methods defined inside the class definition or template functions is not necessary, and at some point we consciously started removing some of them.)